File: php/pro/test/Exchange/test_watch_order_book.php
<?php

function print_orderbook($orderbook, $symbol) {
    $id = isset($orderbook['nonce']) ? $orderbook['nonce'] : $orderbook['datetime'];
    echo $orderbook['datetime'], ' ', $id, ' ', $symbol, ' ',
        count($orderbook['asks']), ' asks ', json_encode($orderbook['asks'][0]), ' ',
        count($orderbook['bids']), ' bids ', json_encode($orderbook['bids'][0]), "\n";
}

function test_watch_order_book($exchange, $symbol) {

    echo __FUNCTION__ . "\n";

    $method = 'watchOrderBook';

    // we have to skip some exchanges here due to the frequency of trading
    $skipped_exchanges = array(
        'ripio',
        'gopax', // requires authentication for watch_order_book
    );

    if (in_array($exchange->id, $skipped_exchanges)) {

        echo $exchange->id, ' ', $method, "() test skipped\n";
        return;
    }

    if (!(array_key_exists($method, $exchange->has) && $exchange->has[$method])) {
        echo $exchange->id, ' ', $method, "() is not supported or not implemented yet\n";
        return;
    }

    $now = $exchange->milliseconds();
    $ends = $now + 10000;
    while ($now < $ends) {
        try {
            $result = yield $exchange->watch_order_book($symbol);
            print_orderbook($result, $symbol);
        } catch (Exception $e) {
            if (!($e instanceof \ccxt\NetworkError)) {
                throw $e;
            }
        }
        $now = $exchange->milliseconds();
    }
}

File: python/ccxt/pro/test/exchange/test_watch_order_book.py
# -*- coding: utf-8 -*-

from exchange.test_order_book import test_order_book
from ccxt import NetworkError


async def test_watch_order_book(exchange, symbol):
    # log (symbol.green, 'watching order book...')
    method = 'watchOrderBook'
    # we have to skip some exchanges here due to the frequency of trading or for other reasons
    skipped_exchanges = [
        'cex',    # CEX requires authentication for watch_order_book
        'ripio',
        'gopax',  # requires authentication for watch_order_book
        'woo',
        'luno',  # requires authentication for watch_order_book
        'alpaca' # requires auth
    ]
    if exchange.id in skipped_exchanges:
        print(exchange.id, method, 'test skipped')
        return
    if (method in exchange.has) and exchange.has[method]:
        response = None
        now = exchange.milliseconds()
        end = now + 10000
        i = 0
        while now < end:
            try:
                exchange.verbose = True
                response = await getattr(exchange, method)(symbol)
                i += 1
                now = exchange.milliseconds()
                test_order_book(exchange, response, method, symbol)
            except NetworkError:
                now = exchange.milliseconds()
        print('Received', i, 'order book messages')
        return response
    else:
        print(method, 'not supported')


__all__ = ['test_watch_order_book']
